<a href="https://colab.research.google.com/github/Haris-09/Python-Practice/blob/main/Python_Primer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#The order of Operations

In [1]:
5 + 30 * 20

605

For example, in the following equation, the numbers 30 and 20 are multiplied first, and the number 5 is added to their product.

Parentheses can change the order of operations.

This equation is another way of saying, “multiply 30 by 20, and then add 5 to the result.” The result is 605. We can change the order of operations by adding parentheses around the first two numbers, like so:

In [2]:
(5 + 30) * 20

700

The result of this equation is 700 (not 605) because the parentheses tell Python to do the operation in the parentheses first, and then do the operation outside the parentheses.

#Nested Parentheses
Inner and outer Parentheses

In [3]:
((5+30)*20)/10

70.0

In this case, Python evaluates the innermost parentheses first, then the outer ones, and then the final division operator. In other words, this equation is saying, “add 5 to 30, then multiply the result by 20, and divide that result by 10.” Here’s what happens:
• Adding 5 to 30 gives 35.
• Multiplying 35 by 20 gives 700.
• Dividing 700 by 10 gives the final answer of 70.

#Variables
Variable in programming means a place to store information such as numbers, text, list of numbers and text etc.Its like variable is a label for something.To create a variable name Ali, we use an equal sign =

In [4]:
Ali=100

Variable names can be made up of letters, numbers and the underscore character ( _ ), but they cant start with a number. You can use anything from single letters such as (a) to long sentences for variable names. But you cannot use space ( )A short name variable is best 

#Variable exercise

In [5]:
ali_coins=20
fahad_coins=10
abdullah_coins=10

ali_coins+fahad_coins+abdullah_coins

40

#Strings in Python

we create a string by putting quotes around text. one can also use sigle or double quotes

In [6]:
abubakr = 'first caliph!!'
print(abubakr)

first caliph!!


In [7]:
abubakr = "first caliph!!"
print(abubakr)

first caliph!!


#Syntax and Syntax Error

*   Syntax means the arrangement and order of word in a sentence.
*   Syntax error means that you did something in an order python was not expecting or you missed something.


In [8]:
abubakr = 'first caliph!!
print(abubakr)

SyntaxError: ignored

#Strings Problems handling

**Adding second line to your command**


*   You can use 3 single quotes to write a second or third or as many lines as you want to a given command.



In [9]:
abubakr= '''first
Righteous
caliph!!'''
print(abubakr)

first
Righteous
caliph!!


to handle a sentence which contains multiple single or double quotes use 3 single quotes to handle this problem.

In [10]:
silly_string= '''He said, "Aren't can't shouldn't wouldn't."'''
print(silly_string)

He said, "Aren't can't shouldn't wouldn't."


**Adding a backslash  \or escaping**


* \or excapingmeans “Yes, I know I have quotes inside my string, and I want you to ignore them until you see the end quote.

In [11]:
single_quote_str= 'He said, "Aren\'t can\'t shouldn\'t wouldn\'t."'
print (single_quote_str)

He said, "Aren't can't shouldn't wouldn't."


In [12]:
double_quote_str= "He said, \"Aren't can't shouldn't wouldn't.\""
print (double_quote_str)

He said, "Aren't can't shouldn't wouldn't."


#Embedding value in strings

(Embedding values, also referred to as string substitution, is programmer-speak for “inserting values.”) For example, to have Python calculate or store the number of points you scored in a game, and then add it to a sentence like “I scored points,” use %s in the sentence in place of the value, and then tell Python that value, like this:

In [13]:
myscore= 1000
message = 'I scored %s points'
print(message % myscore)

I scored 1000 points


In [14]:
nums= 'What did the number %s say to the number %s? Nice belt!!'
print(nums% (0, 8))

What did the number 0 say to the number 8? Nice belt!!


#Multiplying Strings

In [15]:
print(10 * 'a')

aaaaaaaaaa


In [16]:
spaces = ' ' * 25
print('%s 12 DHA Phase 5' % spaces)
print('%s Clifton' % spaces)
print('%s West Snoring' % spaces)
print()
print()
print('Dear Sir')
print()
print('I wish to report that tiles are missing from the')
print('outside toilet roof.')
print('I think it was bad wind the other night that blew them away.')
print()
print('Regards')
print('Sharfoo')

                          12 DHA Phase 5
                          Clifton
                          West Snoring


Dear Sir

I wish to report that tiles are missing from the
outside toilet roof.
I think it was bad wind the other night that blew them away.

Regards
Sharfoo


#Lists vs. string

**Creating a list**

In [17]:
sharfoo_list=['sabzee, fruit, aloo, chai']
print(sharfoo_list[0])

sabzee, fruit, aloo, chai


Creating a list takes a bit more typing than creating a string, but a list is more useful than a string because it can be manipulated.

In [18]:
sharfoo_list=['sabzee', 'fruit', 'aloo', 'chai']
print(sharfoo_list[2])

aloo


In [19]:
print(sharfoo_list[0:3])

['sabzee', 'fruit', 'aloo']


Writing [0:3] is the same as saying, “show the items from index position 0 up to (but not including) index position 3”—or in other words, items 0, 1, and 2.

Lists can store all sorts of items, like numbers and strings

In [20]:
some_numbers= [1, 2, 5, 10, 20]

In [21]:
some_strings= ['Which', 'Witch', 'Is', 'Which']

List strings and number example

In [22]:
numbers_and_strings= ['Why', 'was', 6, 'afraid', 'of', 7, 'because', 7, 8, 9]
print(numbers_and_strings)

['Why', 'was', 6, 'afraid', 'of', 7, 'because', 7, 8, 9]


List within a list 😀

In [23]:
numbers = [1, 2, 3, 4]
strings = ['I', 'kicked', 'my', 'toe', 'and', 'it', 'is', 'sore']
mylist= [numbers, strings]
print(mylist)

[[1, 2, 3, 4], ['I', 'kicked', 'my', 'toe', 'and', 'it', 'is', 'sore']]


**Adding items to a LIST**

In [24]:
wizard_list= ['spider legs', 'toe of frog', 'eye of newt', 'bat wing', 'slug butter', 'snake dandruff']
wizard_list.append('bear burp')
print(wizard_list)

['spider legs', 'toe of frog', 'eye of newt', 'bat wing', 'slug butter', 'snake dandruff', 'bear burp']


In [25]:
wizard_list.append('mandrake')
wizard_list.append('hemlock')
wizard_list.append('swamp gas')

In [26]:
print(wizard_list)

['spider legs', 'toe of frog', 'eye of newt', 'bat wing', 'slug butter', 'snake dandruff', 'bear burp', 'mandrake', 'hemlock', 'swamp gas']


**Removing items from the list**

In [27]:
wizard_list= ['spider legs', 'toe of frog', 'eye of newt', 'bat wing', 'slug butter', 'snake dandruff']
wizard_list.append('bear burp')
del wizard_list[5]
print(wizard_list)

['spider legs', 'toe of frog', 'eye of newt', 'bat wing', 'slug butter', 'bear burp']


**List Arithmetic**

We can join two lists by using +

In [28]:
list1 = [1, 2, 3, 4]
list2 = ['I', 'ate', 'chocolate', 'and', 'I', 'want', 'more']
list3 = list1 + list2
print(list3)

[1, 2, 3, 4, 'I', 'ate', 'chocolate', 'and', 'I', 'want', 'more']


And we can multiply a list by a number.

In [29]:
list1 = [1, 2]
print(list1 * 5)

[1, 2, 1, 2, 1, 2, 1, 2, 1, 2]


On the other hand, division (/) and subtraction (-) give only errors

In [30]:
list1 = [1, 2]
list1 / 20
print(list1)

TypeError: ignored

In [31]:
list1 -20

TypeError: ignored

#Tuples

A tuple is like a list that uses parentheses, as in this example:

In [32]:
fibs = (0, 1, 1, 2, 3)
print(fibs[3])

2


The main difference between a tuple and a list is that a tuple cannot change once you’ve created it.

In [33]:
fibs[0] = 4

TypeError: ignored

#Python Maps

In Python, a map (also referred to as a dict, short for dictionary) is a collection of things, like lists and tuples. The difference between maps and lists or tuples is that each item in a map has a key and a corresponding value. For example, say we have a list of people and their favorite sports. We could put this information into a Python list, with the person’s name followed by their sport, like so:

**Creating a map in Python**

In [36]:
favorite_sports= {'Ali' : 'Football',
                  'Omer' : 'Basketball',
                  'Osman' : 'Baseball',
                  'Abu Bakr' : 'Netball',
                  'Fatima' : 'Badminton',
                  'Faizan' : 'Rugby'}
print(favorite_sports['Ali'])

Football


**Modifying a value in a map**

In [37]:
favorite_sports['Faizan'] = 'Cricket'
print(favorite_sports['Faizan'])

Cricket


**Deleting a value in the map**

In [38]:
del favorite_sports['Ali']
print(favorite_sports)

{'Omer': 'Basketball', 'Osman': 'Baseball', 'Abu Bakr': 'Netball', 'Fatima': 'Badminton', 'Faizan': 'Cricket'}


As you can see, working with maps is kind of like working with lists and tuples, except that you can’t join maps with the plus operator (+). If you try to do that, you’ll get an error message: